linkerd/stack/src/map_err.rs

```diff
@@ -1,58 +1,263 @@
-use futures::TryFutureExt;
+use crate::{layer, ExtractParam, NewService};
 use linkerd_error::Error;
-use std::task::{Context, Poll};
+use std::{
+    fmt,
+    future::Future,
+    marker::PhantomData,
+    pin::Pin,
+    task::{Context, Poll},
+};
+
+pub trait WrapErr<E> {
+    type Error: Into<Error>;
+
+    fn wrap_err(&self, error: E) -> Self::Error;
+}
+
+impl<E: Into<Error>> WrapErr<E> for () {
+    type Error = Error;
+
+    fn wrap_err(&self, error: E) -> Error {
+        error.into()
+    }
+}
+
+impl<InE, OutE, F> WrapErr<InE> for F
+where
+    F: FnOnce(InE) -> OutE + Clone,
+    OutE: Into<Error>,
+{
+    type Error = OutE;
+
+    fn wrap_err(&self, error: InE) -> OutE {
+        (self.clone())(error)
+    }
+}
+
+/// A [`NewService`] that extracts a [`WrapErr`] implementation from its target
+/// and produces a [`MapErr`] middleware that wrapps inner error types.
+pub struct NewMapErr<W, X, N> {
+    inner: N,
+    extract: X,
+    _cx: PhantomData<fn(W)>,
+}
 
 /// Like `tower::util::MapErr`, but with an implementation of `Proxy`.
 #[derive(Clone, Debug)]
-pub struct MapErr<S, F> {
+pub struct MapErr<W, S> {
     inner: S,
-    f: F,
+    wrap: W,
+}
+
+/// Futures returned by [`AnnotateError`].
```

Comment:
docs need to be updated here
```suggestion
/// Futures returned by [`MapErr`].
```